tests/test_treebandit.py

```diff
@@ -0,0 +1,348 @@
+                                learning_policy=LearningPolicy.UCB1(),
+                                neighborhood_policy=NeighborhoodPolicy.TreeBandit(tree_parameters=tree_parameters),
+                                context_history=[[0, 1, 2, 3], [1, 2, 3, 0], [2, 3, 1, 0], [3, 2, 1, 0]],
+                                contexts=[[2, 3, 1, 0]],
+                                seed=123456,
+                                num_run=1,
+                                is_predict=True)
+
+        self.assertEqual(mab._imp.arm_to_tree[arm].__dict__["criterion"], "entropy")
+        self.assertEqual(mab._imp.arm_to_tree[arm].__dict__["max_depth"], 4)
+        self.assertEqual(mab._imp.arm_to_tree[arm].__dict__["min_samples_split"], 2)
+        self.assertEqual(mab._imp.arm_to_tree[arm].__dict__["max_leaf_nodes"], 10)
+
+    def test_greedy0(self):
+        arms, mab = self.predict(arms=[1, 2, 3, 4],
+                                 decisions=[1, 1, 1, 2, 2, 3, 3, 3, 3, 3],
+                                 rewards=[0, 1, 1, 0, 0, 0, 0, 1, 1, 1],
+                                 learning_policy=LearningPolicy.EpsilonGreedy(epsilon=0),
+                                 neighborhood_policy=NeighborhoodPolicy.TreeBandit(),
+                                 context_history=[[0, 1, 2, 3, 5], [1, 1, 1, 1, 1], [0, 0, 1, 0, 0],
+                                                  [0, 2, 2, 3, 5], [1, 3, 1, 1, 1], [0, 0, 0, 0, 0],
+                                                  [0, 1, 4, 3, 5], [0, 1, 2, 4, 5], [1, 2, 1, 1, 3],
+                                                  [0, 2, 1, 0, 0]],
+                                 contexts=[[0, 1, 2, 3, 5], [1, 1, 1, 1, 1]],
+                                 seed=123456,
+                                 num_run=1,
+                                 is_predict=True)
+
+        self.assertListEqual(arms, [1, 1])
+
+    def test_greedy0_single(self):
+        arm, mab = self.predict(arms=[1, 2, 3, 4],
+                                decisions=[1, 1, 1, 2, 2, 3, 3, 3, 3, 3],
+                                rewards=[0, 1, 1, 0, 0, 0, 0, 1, 1, 1],
+                                learning_policy=LearningPolicy.EpsilonGreedy(epsilon=0),
+                                neighborhood_policy=NeighborhoodPolicy.TreeBandit(),
+                                context_history=[[0, 1, 2, 3, 5], [1, 1, 1, 1, 1], [0, 0, 1, 0, 0],
+                                                 [0, 2, 2, 3, 5], [1, 3, 1, 1, 1], [0, 0, 0, 0, 0],
+                                                 [0, 1, 4, 3, 5], [0, 1, 2, 4, 5], [1, 2, 1, 1, 3],
+                                                 [0, 2, 1, 0, 0]],
+                                contexts=[[0, 1, 2, 3, 5]],
+                                seed=123456,
+                                num_run=1,
+                                is_predict=True)
+
+        self.assertEqual(arm, 1)
+
+    def test_greedy1(self):
+        arms, mab = self.predict(arms=[1, 2, 3, 4],
```

Comment:
@irmakbky curious how you came up with what values to assert. Pen and paper? These tests seem to be a bit more complicated from a dummy set to just "see " what it should equal to.

Reply:
With learning policies like greedy with epsilon=0, since the context we're trying to predict for has been observed before, we know that in the trees, we'll end up reaching the same leaves. So, we expect it to predict the arm that gave the highest reward in the training. Although, now looking at it, maybe I should add more training data for the same context to test among the arm predictions better!

For other learning policies with exploration, I've had to observe the prediction first and then assert it. So, we're only checking if it predicts the same thing each time.

Reply:
excellent!

Reply:
In this test epsilon=1, ie there should be 100% randomization. I'm a little concerned that it's giving identical results to when epsilon=0. Could you add a second test with a different random seed to make sure that there isn't something else going on?

Reply:
(increasing the number of test contexts would would also work, eg predict 4 rather than 2)

Reply:
Makes sense! I've made it so that test_greedy0 and test_greedy1 have the same example but guess different arms for the first context. Is this ok?